### PR TITLE
DWARF 4 and 5 line parsing additions ##bin

### PR DESCRIPTION
**Pull request by HoundThe**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

I am fixing parsing for DWARF 4 and 5 line information and going to add any extra changes introduced by the versions (especially 5, because it did a lot of changes).

I also did small refactorings, like moving the function prototypes out of the r_bin.h and moved it to the r_bin_dwarf.h etc.

### CODE REVIEW

File: libr/bin/dwarf.c [hunk 1]
```diff
@@ -396,7 +339,132 @@ static const ut8 *r_bin_dwarf_parse_lnp_header(
...
+// 	sdb_free (sdb);
+
+	return NULL;
+}
+
+static const ut8 *parse_line_header (
+	RBinFile *bf, const ut8 *buf, const ut8 *buf_end,
+	RBinDwarfLineHeader *hdr, FILE *f, int mode) {
+
+	if (!hdr || !bf || !buf) {
```
Comment (trufae) [inline libr/bin/dwarf.c:377]: use r_return_val_if_fail

File: libr/bin/dwarf.c [hunk 2]
```diff
@@ -330,26 +262,37 @@ static const ut8 *r_bin_dwarf_parse_lnp_header(
...
 				if (id_idx > 0) {
-					include_dir = pinclude_dir = sdb_array_get (s, "includedirs", id_idx - 1, 0);
+					include_dir = pinclude_dir = sdb_array_get (sdb, "includedirs", id_idx - 1, 0);
 					if (include_dir && include_dir[0] != '/') {
 						comp_dir = sdb_get (bf->sdb_addrinfo, "DW_AT_comp_dir", 0);
 						if (comp_dir) {
-							allocated_id = calloc (1, strlen (comp_dir) +
-									strlen (include_dir) + 8);
+							allocated_id = calloc (1, strlen (comp_dir) + strlen (include_dir) + 8);
 							snprintf (allocated_id, strlen (comp_dir) + strlen (include_dir) + 8,
```
Comment (trufae) [inline libr/bin/dwarf.c:294]: use r_str_newf instead of calloc+snprintf

File: libr/bin/dwarf.c [hunk 3]
```diff
@@ -213,98 +213,29 @@ static void r_bin_dwarf_header_fini(RBinDwarfLNPHeader *hdr) {
...
-	i = 0;
-	while (buf+1 < buf_end) {
-		int maxlen = R_MIN ((size_t)(buf_end-buf)-1, 0xfff);
-		int len = r_str_nlen ((const char*)buf, maxlen);
+	while (buf + 1 < buf_end) {
+		int maxlen = R_MIN ((size_t) (buf_end - buf) - 1, 0xfff);
+		int len = r_str_nlen ((const char *)buf, maxlen);
 		char *str = r_str_ndup ((const char *)buf, len);
-		if (len<1 || len >= 0xfff) {
+		if (len < 1 || len >= 0xfff) {
```
Comment (trufae) [inline libr/bin/dwarf.c:230]: 0xfff seems like an arbitrary limit

File: libr/bin/dwarf.c [hunk 4]
```diff
@@ -213,98 +213,29 @@ static void r_bin_dwarf_header_fini(RBinDwarfLNPHeader *hdr) {
...
-	} else {
-		hdr->std_opcode_lengths = NULL;
-	}
-
-	i = 0;
-	while (buf+1 < buf_end) {
-		int maxlen = R_MIN ((size_t)(buf_end-buf)-1, 0xfff);
-		int len = r_str_nlen ((const char*)buf, maxlen);
+	while (buf + 1 < buf_end) {
+		int maxlen = R_MIN ((size_t) (buf_end - buf) - 1, 0xfff);
```
Comment (trufae) [inline libr/bin/dwarf.c:227]: we can use size_t

File: libr/bin/dwarf.c (hunk 3 above)
Comment (HoundThe) [reply libr/bin/dwarf.c:230]: Yeah, but I guess it it's sufficient for any real dir length? And it's somewhat of an obscure protection against malformed DWARF. Not sure what the original writer intention was tho.

**Comment by HoundThe** [comment]
> code looks good, i'm missing a test
Tests are here, waiting for this to be merged
https://github.com/radareorg/radare2/pull/16965

File: libr/bin/dwarf.c [hunk 5]
```diff
@@ -330,27 +262,34 @@ static const ut8 *r_bin_dwarf_parse_lnp_header(
...
 			}
 			buf += len + 1;
-			if (buf>=buf_end) { buf = NULL; goto beach; }
-			buf = r_uleb128 (buf, buf_end-buf, &id_idx);
-			if (buf>=buf_end) { buf = NULL; goto beach; }
-			buf = r_uleb128 (buf, buf_end-buf, &mod_time);
-			if (buf>=buf_end) { buf = NULL; goto beach; }
-			buf = r_uleb128 (buf, buf_end-buf, &file_len);
-			if (buf>=buf_end) { buf = NULL; goto beach; }
+			if (buf >= buf_end) {
```
Comment (XVilka) [inline libr/bin/dwarf.c:265]: This could be simplified with a macro in the future @HoundThe, in the subsequent pull requests.

**Pull request merged**